src/netlink.c

```diff
@@ -0,0 +1,431 @@
+ * netlink transactions for a given thread.
+ *
+ * The handle is not thread safe
+ *
+ * \return \c handle to netlink on success, else \c NULL.
+ */
+struct netlink_handle *
+netlink_init(void) {
+	struct netlink_handle *hndl = g_malloc0(sizeof(*hndl));
+
+	hndl->nl = mnl_socket_open(NETLINK_ROUTE);
+	if (hndl->nl == NULL) {
+		g_debug("mnl_socket_open %s", strerror(errno));
+		goto out;
+	}
+
+	if (mnl_socket_bind(hndl->nl, 0, MNL_SOCKET_AUTOPID) < 0) {
+		g_debug("mnl_socket_bind %s", strerror(errno));
+		netlink_close(hndl);
+		goto out;
+	}
+
+	hndl->seq = (guint)time(NULL);
+	return hndl;
+
+out:
+	g_free(hndl);
+	return NULL;
+}
+
+/*!
+ * Close the netlink connection
+ *
+ * \param \c handle returned from a call to netlink_init
+ */
+void
+netlink_close(struct netlink_handle *const hndl) {
+	if (hndl == NULL) {
+		return;
+	}
+
+	if (mnl_socket_close(hndl->nl) == -1) {
+		g_debug("mnl_socket_close %s", strerror(errno));
+	}
+}
+
+/*!
+ * Execute a netlink transaction and check the result
+ * This method can be used for any transaction where
+ * only the success or failure of the transaction needs
+ * to be known (i.e. no data is send back)
+ *
+ * \param handle returned from a call to netlink_init
+ * \param nlh pre created netlink message
+ *
+ * \return \c true on success else \c false
+ */
+static gboolean
+netlink_execute(struct netlink_handle *const hndl,
+		struct nlmsghdr *const nlh) {
+	gchar buf[MNL_SOCKET_BUFFER_SIZE];
+	ssize_t ret = -1;
+	gboolean status = false;
+	guint  portid, seq;
+
+	if (hndl == NULL) {
+		g_debug("netlink_link_execute: NULL pointer");
+		return false;
+	}
+
+	if (hndl->nl == NULL) {
+		g_debug("netlink_link_execute: NULL pointer");
+		return false;
+	}
+
+	nlh->nlmsg_seq = seq = hndl->seq++;
+	portid = mnl_socket_get_portid(hndl->nl);
+
+	if (mnl_socket_sendto(hndl->nl, nlh, nlh->nlmsg_len) < 0) {
+		g_debug("mnl_socket_sendto %s", strerror(errno));
+		goto out;
+	}
+
+	ret = mnl_socket_recvfrom(hndl->nl, buf, sizeof(buf));
+	if (ret == -1) {
+		g_debug("mnl_socket_recvfrom failed %s", strerror(errno));
+		goto out;
+	}
+
+	while (ret > 0) {
+		ret = mnl_cb_run(buf, (size_t)ret, seq, portid, NULL, NULL);
+		if (ret <= 0) {
+			break;
+		}
+		ret = mnl_socket_recvfrom(hndl->nl, buf, sizeof(buf));
+	}
+
+	if (ret == -1) {
+		g_debug("netlink error");
+		goto out;
+	}
+
+	status = true;
+out:
+	return status;
+}
+
+/*!
+ * Netlink command equivalent to
+ * ip link set dev <interface> <up|down> 
+ *
+ * \param handle returned from a call to netlink_init
+ * \param interface device name to enable\disable
+ * \param enable if true device will enabled, else disabled
+ *
+ * \return \c true on success else \c false
+ */
+gboolean
+netlink_link_enable(struct netlink_handle *const hndl,
+		    const gchar *const interface, gboolean enable)  {
+	char buf[MNL_SOCKET_BUFFER_SIZE];
+	struct nlmsghdr *nlh;
+	struct ifinfomsg *ifm;
+	guint change = 0, flags = 0;
+
+	g_debug("netlink_link_enable %s", interface);
```

Comment:
Please could you add a `%d` for the `enable` param so it's clear in the log which operation the function is performing.